crates/codegen/parser/runtime/src/query/engine.rs

```diff
@@ -0,0 +1,481 @@
+use std::collections::HashMap;
+use std::rc::Rc;
+
+// This crate is copied to another crate, so all imports should be relative
+use super::super::cst;
+use super::super::cursor::Cursor;
+use super::model::{
+    AlternativesMatcher, BindingMatcher, Kind, Matcher, NodeMatcher, NodeSelector,
+    OneOrMoreMatcher, OptionalMatcher, Query, SequenceMatcher,
+};
+
+impl Cursor {
+    pub fn query(self, queries: Vec<Query>) -> QueryResultIterator {
+        QueryResultIterator::new(self, queries)
+    }
+
+    fn consume(&mut self) -> bool {
+        if self.is_completed() {
+            false
+        } else {
+            if !self.go_to_next_sibling() {
+                self.complete();
+            }
+            true
+        }
+    }
+
+    fn matches_node_selector(&self, node_selectot: &NodeSelector) -> bool {
+        match self.node() {
+            cst::Node::Rule(rule) => match node_selectot {
+                NodeSelector::Anonymous => true,
+                NodeSelector::Kind { kind } => Kind::Rule(rule.kind) == *kind,
+                NodeSelector::Text { .. } => false,
+                NodeSelector::FieldName { field_name: field } => Some(*field) == self.node_name(),
```

Comment:
nit: leftover from the rename:

.. also others ..

```suggestion
                NodeSelector::FieldName { field_name } => Some(*field_name) == self.node_name(),
```

Reply:
Done